lib/Conversion/HandshakeToHW/HandshakeToHW.cpp

```diff
@@ -272,225 +317,422 @@ static Operation *checkSubModuleOp(mlir::ModuleOp parentModule,
-                     rewriter.getI1Type(), inIdx, StringAttr{}});
+    ports.inputs.push_back({rewriter.getStringAttr("clock"),
+                            PortDirection::INPUT, rewriter.getI1Type(), inIdx++,
+                            StringAttr{}});
+    ports.inputs.push_back({rewriter.getStringAttr("reset"),
+                            PortDirection::INPUT, rewriter.getI1Type(), inIdx,
+                            StringAttr{}});
   }
 
   return ports;
 }
 
 /// Returns a vector of PortInfo's which defines the FIRRTL interface of the
 /// to-be-converted op.
-static llvm::SmallVector<PortInfo>
-getPortInfoForOp(ConversionPatternRewriter &rewriter, Operation *op) {
+static ModulePortInfo getPortInfoForOp(ConversionPatternRewriter &rewriter,
+                                       Operation *op) {
   return getPortInfoForOp(rewriter, op, op->getOperandTypes(),
                           op->getResultTypes());
 }
 
-/// All standard expressions and handshake elastic components will be converted
-/// to a HW sub-module and be instantiated in the top-module.
-static HWModuleExternOp createSubModuleOp(ModuleOp parentModule,
-                                          Operation *oldOp,
-                                          ConversionPatternRewriter &rewriter) {
-  OpBuilder::InsertionGuard g(rewriter);
-  rewriter.setInsertionPointToStart(parentModule.getBody());
-  auto ports = getPortInfoForOp(rewriter, oldOp);
-  // todo: HWModuleOp; for this initial commit we'll leave this as an extern op.
-  return rewriter.create<HWModuleExternOp>(
-      parentModule.getLoc(), rewriter.getStringAttr(getSubModuleName(oldOp)),
-      ports);
-}
+namespace {
 
-//===----------------------------------------------------------------------===//
-// HW Top-module Related Functions
-//===----------------------------------------------------------------------===//
+// Input handshakes contain a resolved valid and (optional )data signal, and
+// a to-be-assigned ready signal.
+struct InputHandshake {
+  Value valid;
+  std::shared_ptr<Backedge> ready;
+  Value data;
+};
 
-static hw::HWModuleOp createTopModuleOp(handshake::FuncOp funcOp,
-                                        ConversionPatternRewriter &rewriter) {
-  llvm::SmallVector<PortInfo, 8> ports = getPortInfoForOp(
-      rewriter, funcOp, funcOp.getArgumentTypes(), funcOp.getResultTypes());
+// Output handshakes contain a resolved ready, and to-be-assigned valid and
+// (optional) data signals.
+struct OutputHandshake {
+  std::shared_ptr<Backedge> valid;
+  Value ready;
+  std::shared_ptr<Backedge> data;
+};
 
-  // Create a HW module.
-  auto hwModuleOp = rewriter.create<hw::HWModuleOp>(
-      funcOp.getLoc(), rewriter.getStringAttr(funcOp.getName()), ports);
+struct UnwrappedIO {
+  llvm::SmallVector<InputHandshake> inputs;
+  llvm::SmallVector<OutputHandshake> outputs;
+
+#define IOAccessor(Type, container, member, name)                              \
```

Comment:
Could this be a template? I think so, but I'm not C++ expert. I do know that we have encouraged templates over macros as a matter of style in the past, but this isn't LLVM policy, so I don't have strong opinions.